### PR TITLE
Add inherent unchecked_shl, unchecked_shr to integers

### PR DESCRIPTION
**Pull request by clarfonthey**
Tracking issue: #85122.

Adding more of these methods, since these are missing.

### CODE REVIEW

**Comment by scottmcm** [comment]
Thanks for the PR, @clarfonthey! 

The shift semantics for Rust are a bit weird, though.  Do you know any situations where the masking the shift amount is a problem in practice?

I feel like someone reaching for these would often actually want things like [`shl n[us]w`](https://llvm.org/docs/LangRef.html#shl-instruction) or `[la]shr exact`, rather than what these intrinsics actually do.

But at the same time I guess we're stuck with `checked_shl` looking only at the shift amount, so maybe this is the only reasonable behaviour?  And then `shl_nowrap` and `shr_exact` would go with `div_exact` in the "well there's another category of these methods that needs a different naming pattern"?

Any thoughts on whether these should be added, @rust-lang/libs ?

**Comment by scottmcm** [comment]
I thought about this some more, and came around to the idea that it's probably fine to have these in nightly because of the parallel with `checked_sh[lr]`.  I'll let libs make the harder decision about stabilization later.

So @clarfonthey can you please:
- Update the documentation to explicitly mention that this is *not* about overflow in output values.  (Not that the docs hint that it could be, but for `unsafe` methods I'd rather be extra clear.)
- Since they're `unsafe`, add a `# Safety` section ([example](https://doc.rust-lang.org/nightly/std/ptr/fn.swap_nonoverlapping.html#safety)) with a bullet for each condition that must be uphelp to avoid being UB.  (You have it in the doc comment already, but to make it easier on the person wanting to write a `// SAFETY` justification on their `unsafe` block, I like seeing it put under a very explicit header.)

**Comment by clarfonthey** [comment]
Will update the docs for the methods in a bit. Also will add the `# Safety` notice to the other unsafe methods as they do not include it either, and I just copied their docs.

File: library/core/src/num/int_macros.rs [hunk 1]
```diff
@@ -645,6 +645,23 @@ macro_rules! int_impl {
             if unlikely!(b) {None} else {Some(a)}
         }
 
+        /// Unchecked shift left. Computes `self << rhs`, resulting in
+        /// undefined behavior `rhs` is larger than or equal to the number of bits in `self`.
```
Comment (CraftSpider) [inline library/core/src/num/int_macros.rs:649]:
```suggestion
        /// undefined behavior if `rhs` is greater than or equal to the number of bits in `self`.
```

File: library/core/src/num/uint_macros.rs [hunk 2]
```diff
@@ -655,6 +655,23 @@ macro_rules! uint_impl {
             if unlikely!(b) {None} else {Some(a)}
         }
 
+        /// Unchecked shift left. Computes `self << rhs`, resulting in
+        /// undefined behavior `rhs` is larger than or equal to the number of bits in `self`.
```
Comment (CraftSpider) [inline library/core/src/num/uint_macros.rs:659]:
```suggestion
        /// undefined behavior if `rhs` is greater than or equal to the number of bits in `self`.
```

File: library/core/src/num/uint_macros.rs [hunk 3]
```diff
@@ -676,6 +693,23 @@ macro_rules! uint_impl {
             if unlikely!(b) {None} else {Some(a)}
         }
 
+        /// Unchecked shift right. Computes `self >> rhs`, resulting in
+        /// undefined behavior `rhs` is largerthan or equal to the number of bits in `self`.
```
Comment (CraftSpider) [inline library/core/src/num/uint_macros.rs:697]:
```suggestion
        /// undefined behavior if `rhs` is greater than or equal to the number of bits in `self`.
```

File: library/core/src/num/int_macros.rs [hunk 4]
```diff
@@ -666,6 +683,23 @@ macro_rules! int_impl {
             if unlikely!(b) {None} else {Some(a)}
         }
 
+        /// Unchecked shift right. Computes `self >> rhs`, resulting in
+        /// undefined behavior `rhs` is largerthan or equal to the number of bits in `self`.
```
Comment (CraftSpider) [inline library/core/src/num/int_macros.rs:687]:
```suggestion
        /// undefined behavior if `rhs` is greater than or equal to the number of bits in `self`.
```

**Comment by CraftSpider** [comment]
Wording nit, not super important as you're already updating the docs, but noticed as I was reading.

**Comment by clarfonthey** [comment]
> Wording nit, not super important as you're already updating the docs, but noticed as I was reading.

Apologies, I just was about to push my changes that fixed that when you commented. :P 

I chose to use "larger than" instead of "greater than" since that's the wording that the existing `checked` methods use.

**Comment by scottmcm** [comment]
Thanks for the updates, @clarfonthey!  I've updated the tracking issue with some extra things to ponder before stabilization.

@bors r+

**Pull request merged**